File: cpp/cpp_1000-/1289.cc
// C++
// leetcode 1289
// https://leetcode-cn.com/problems/minimum-falling-path-sum-ii/

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& grid) {
        int n = grid.size();
        vector<int> dp(grid[0]);
        for(int i = 1; i < n; i++) {
            int min1 = INT_MAX;
            int min2 = INT_MAX;
            int min1pos = -1;
            for(int j = 0; j < n; j++) {
                if(dp[j] <= min1) {
                    min2 = min1;
                    min1 = dp[j];
                    min1pos = j;
                } else if(dp[j] < min2) {
                    min2 = dp[j];
                }
            }
            for(int j = 0; j < n; j++) {
                if(j == min1pos) {
                    dp[j] = grid[i][j] + min2;
                } else {
                    dp[j] = grid[i][j] + min1;
                }
            }
        }
        int res = INT_MAX;
        for(auto x : dp)
            res = min(res, x);
        return  res;
    }
};

File: Go/Go_1000-/1289.go
// Go
// leetcode 1289
// https://leetcode.cn/problems/minimum-falling-path-sum-ii/

func minFallingPathSum(grid [][]int) int {
	n := len(grid)
	arr := make([]int, n)
	for i:=0; i<n; i++ {
		arr[i] = grid[0][i]
	}
	mn1, mn2 := 100, 100
	update := func(x int) {
		mn2 = min(mn2, x)
		if mn1 > mn2 {
			mn1, mn2 = mn2, mn1
		}
	}
	tmp := make([]int, n)
	for row := 1; row < n; row++ {
		mn1, mn2 = 30000, 30000
		for _, x := range arr {
			update(x)
		}
		for i:=0; i<n; i++ {
			if arr[i] == mn1 {
				tmp[i] = mn2 + grid[row][i]
			} else {
				tmp[i] = mn1 + grid[row][i]
			}
		}

		for i:=0; i<n; i++ {
			arr[i] = tmp[i]
		}
	}

	res := 30000
	for _, x := range arr {
		res = min(res, x)
	}

	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}